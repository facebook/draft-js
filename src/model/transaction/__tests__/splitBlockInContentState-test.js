/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 */

'use strict';

jest.autoMockOff();

const Immutable = require('immutable');
const getSampleStateForTesting = require('getSampleStateForTesting');
const splitBlockInContentState = require('splitBlockInContentState');

describe('removeRangeFromContentState', () => {
  const {
    contentState,
    selectionState,
  } = getSampleStateForTesting();

  function checkForCharacterList(block) {
    expect(Immutable.List.isList(block.getCharacterList())).toBe(true);
  }

  function getInlineStyles(block) {
    return block.getCharacterList().map(c => c.getStyle()).toJS();
  }

  function getEntities(block) {
    return block.getCharacterList().map(c => c.getEntity()).toJS();
  }

  it('must be restricted to collapsed selections', () => {
    expect(() => {
      const nonCollapsed = selectionState.set('focusOffset', 1);
      return splitBlockInContentState(contentState, nonCollapsed);
    }).toThrow();

    expect(() => {
      return splitBlockInContentState(contentState, selectionState);
    }).not.toThrow();
  });

  it('must split at the beginning of a block', () => {
    const initialBlock = contentState.getBlockMap().first();
    const afterSplit = splitBlockInContentState(contentState, selectionState);
    const afterBlockMap = afterSplit.getBlockMap();
    expect(afterBlockMap.size).toBe(4);

    const preSplitBlock = afterBlockMap.first();

    expect(preSplitBlock.getKey()).toBe(initialBlock.getKey());
    expect(preSplitBlock.getText()).toBe('');
    expect(getInlineStyles(preSplitBlock)).toEqual([]);
    expect(getEntities(preSplitBlock)).toEqual([]);

    const postSplitBlock = afterBlockMap.skip(1).first();
    expect(preSplitBlock.getKey()).not.toBe(postSplitBlock.getKey());
    expect(preSplitBlock.getType()).toBe(postSplitBlock.getType());

    expect(postSplitBlock.getKey()).not.toBe(initialBlock.getKey());
    expect(postSplitBlock.getType()).toBe(initialBlock.getType());
    expect(postSplitBlock.getText()).toBe(initialBlock.getText());
    expect(
      getInlineStyles(initialBlock)
    ).toEqual(
      getInlineStyles(postSplitBlock)
    );
    expect(
      getEntities(postSplitBlock)
    ).toEqual(
      getEntities(initialBlock)
    );

    checkForCharacterList(preSplitBlock);
    checkForCharacterList(postSplitBlock);
  });

  it('must split within a block', () => {
    const initialBlock = contentState.getBlockMap().first();
    const SPLIT_OFFSET = 3;
    const selection = selectionState.merge({
      anchorOffset: SPLIT_OFFSET,
      focusOffset: SPLIT_OFFSET,
    });

    const afterSplit = splitBlockInContentState(contentState, selection);
    const afterBlockMap = afterSplit.getBlockMap();
    expect(afterBlockMap.size).toBe(4);

    const preSplitBlock = afterBlockMap.first();
    const postSplitBlock = afterBlockMap.skip(1).first();

    expect(preSplitBlock.getKey()).toBe(initialBlock.getKey());
    expect(preSplitBlock.getText()).toBe(
      initialBlock.getText().slice(0, SPLIT_OFFSET)
    );
    expect(getInlineStyles(preSplitBlock)).toEqual(
      getInlineStyles(initialBlock).slice(0, SPLIT_OFFSET)
    );
    expect(getEntities(preSplitBlock)).toEqual(
      getEntities(initialBlock).slice(0, SPLIT_OFFSET)
    );

    expect(preSplitBlock.getKey()).not.toBe(postSplitBlock.getKey());
    expect(preSplitBlock.getType()).toBe(postSplitBlock.getType());

    expect(postSplitBlock.getKey()).not.toBe(initialBlock.getKey());
    expect(postSplitBlock.getType()).toBe(initialBlock.getType());
    expect(postSplitBlock.getText()).toBe(
      initialBlock.getText().slice(SPLIT_OFFSET)
    );
    expect(getInlineStyles(postSplitBlock)).toEqual(
      getInlineStyles(initialBlock).slice(SPLIT_OFFSET)
    );
    expect(getEntities(postSplitBlock)).toEqual(
      getEntities(initialBlock).slice(SPLIT_OFFSET)
    );

    checkForCharacterList(preSplitBlock);
    checkForCharacterList(postSplitBlock);
  });

  it('must split at the end of a block', () => {
    const initialBlock = contentState.getBlockMap().first();
    const end = initialBlock.getLength();
    const selection = selectionState.merge({
      anchorOffset: end,
      focusOffset: end,
    });

    const afterSplit = splitBlockInContentState(contentState, selection);
    const afterBlockMap = afterSplit.getBlockMap();
    expect(afterBlockMap.size).toBe(4);

    const preSplitBlock = afterBlockMap.first();
    const postSplitBlock = afterBlockMap.skip(1).first();

    expect(preSplitBlock.getKey()).toBe(initialBlock.getKey());

    expect(preSplitBlock.getKey()).not.toBe(postSplitBlock.getKey());
    expect(preSplitBlock.getType()).toBe(postSplitBlock.getType());
    expect(preSplitBlock.getText()).toBe(initialBlock.getText());
    expect(
      getInlineStyles(preSplitBlock)
    ).toEqual(
      getInlineStyles(initialBlock)
    );
    expect(
      getEntities(preSplitBlock)
    ).toEqual(
      getEntities(initialBlock)
    );

    expect(postSplitBlock.getKey()).not.toBe(initialBlock.getKey());
    expect(postSplitBlock.getType()).toBe(initialBlock.getType());
    expect(postSplitBlock.getText()).toBe('');
    expect(getInlineStyles(postSplitBlock)).toEqual([]);
    expect(getEntities(postSplitBlock)).toEqual([]);

    checkForCharacterList(preSplitBlock);
    checkForCharacterList(postSplitBlock);
  });
});
