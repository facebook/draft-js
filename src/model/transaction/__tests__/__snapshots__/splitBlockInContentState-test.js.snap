// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must split at the beginning of a block 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "key1": Immutable.Record {
    "key": "key1",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must split at the end of a block 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "key3": Immutable.Record {
    "key": "key3",
    "type": "unstyled",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must split within a block 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alp",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "key2": Immutable.Record {
    "key": "key2",
    "type": "unstyled",
    "text": "ha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;
